server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -412,11 +417,68 @@ public void executeQueryPhase(ShardSearchRequest request, SearchShardTask task,
                     return;
                 }
             }
-            // fork the execution in the search thread pool
-            runAsync(getExecutor(shard), () -> executeQueryPhase(orig, task), l);
+            ensureAfterSeqNoRefreshed(shard, orig, () -> executeQueryPhase(orig, task), l);
         }));
     }
 
+    private <T> void ensureAfterSeqNoRefreshed(IndexShard shard, ShardSearchRequest request, CheckedSupplier<T, Exception> executable,
+                                               ActionListener<T> listener) {
+        final Executor executor = getExecutor(shard);
+        executor.execute(new ActionRunnable<>(listener) {
```

Comment:
I am not sure if we should fork at this point. Can we execute in the network thread directly since the refresh is async (if we disable the force refresh when listeners are full) ?

Reply:
I'm not sure why we are opposed to forking here. This is the point where the prior functionality forked I believe. I understand it would be ideal to not do this, but I feel like it is more complicated to wade into changed the behavior around RefreshListeners in this initial PR.

The fork point is pretty much identical for normal search requests. The wait_for_checkpoint search requests could theoretically fork later (after some validation, refresh listener, and timeout scheduling). But as I have mentioned I feel like addressing forced refreshes and stuff like that would be better focused on in a follow-up.

Reply:
I don't like the fork because it relies on another listener so now we require to enter the search thread pool twice ?